zebra: clean up VRF handling by using dataplane provided vrf_id

maxime-leroy opened the pull request:
Zebra was implicitly assuming that the VRF netdevice ifindex could be used as the VRF identifier, leading to casts from ifindex to vrf_id_t. While this works with the Linux kernel dataplane, it does not respect the dataplane API, which exposes these identifiers as distinct concepts.

The code is updated to use the VRF identifier explicitly provided by the dataplane context and removes the need for casting ifindex to vrf_id_t. In addition, the VRF delete path is clarified by no longer reading unused dataplane fields, avoiding confusion about which identifiers are actually required.

There is no behavior change for the Linux kernel dataplane.

vjardin commented on zebra/interface.c:
```diff
@@ -1482,89 +1482,83 @@ static void zebra_if_netconf_update_ctx(struct zebra_dplane_ctx *ctx,
-			zlog_debug(
-				"DPLANE_OP_INTF_UPDATE for VRF %s(%u) table %u",
-				name, ifindex, tableid);
+	frrtrace(2, frr_zebra, if_vrf_del, vrf_id, name);
```
I do not see any if_vrf_del from zebra_trace.h but if_vrf_change was defined.

What's about tools/frr_babeltrace.py ?

vjardin commented on zebra/interface.c:
```diff
@@ -1992,16 +1986,16 @@ static void zebra_if_dplane_ifp_handling(struct zebra_dplane_ctx *ctx)
 		else if (IS_ZEBRA_IF_VXLAN(ifp))
 			zebra_l2_vxlanif_del(ifp);
 
+		vrf_id = ifp->vrf->vrf_id;
```
can we trust that ifp->vrf is neve NULL ?

vjardin commented on zebra/interface.c:
```diff
@@ -1482,89 +1482,83 @@ static void zebra_if_netconf_update_ctx(struct zebra_dplane_ctx *ctx,
+		vrf = vrf_lookup_by_id(exist_id);
+
+		if (!vrf_lookup_by_id(vrf_id) && !vrf) {
+			flog_err(EC_ZEBRA_VRF_NOT_FOUND, "VRF %s id %u does not exist", name,
```
ok, it makes sense.

vjardin reviewed (approved):
thanks, LGTM

mjstapp reviewed (commented):
there's some confusion here - but I'm not sure that this PR isn't making things harder to understand.
I agree that it's confusing to repeatedly cast a value that appears as an ifindex - that's not a great pattern. but I can imagine a very small couple of lines that would correct that in interface.c - I don't see that reorganizing the whole vrf device change flow is helping.

maxime-leroy commented:
> yes, I understood the two commits, I think. my point was: your choice to "clean up" what is the common handler function had a significant impact. you could just have ensured that the vrf_id value was conveyed. but instead, you removed the handler, replaced it with two new functions, encountered the lttng traces that are involved. a two-line change turned into 100 lines. that has an impact on the reviewers as well - we have to try to confirm that your changes retain the important logic that you propose moving into separate functions, we have to sort out the traceing api issues. so: I think you should not try to make that "clean up" change, and just get the vrf_id straight.
> 

Thanks for the detailed feedback — I understand the concern about review surface / churn.

That said, I’d like to clarify the intent and why I still think the split is justified:

- The commits are intentionally split:

    The first commit is cleanup-only, and the second commit contains the actual fix (using the dplane-provided vrf_id). 
    This is deliberate to separate “mechanical refactor” from “functional change” and make the review easier.

- No functional change is introduced by the first commit:

   This is explicitly stated in the commit log and PR description, and I also mentioned during the weekly meeting that the first commit was only a split/cleanup with no functional change. The Linux kernel dataplane behavior is unchanged.

- The cleanup is not only cosmetic:

    In the netlink backend, ifp_table_id is set for RTM_NEWLINK VRF events but not for RTM_DELLINK. The delete path does not require table_id, and reading/tracing it on delete can produce misleading trace data (e.g., a delete event always carrying table_id == 0). Splitting the flow makes the required inputs explicit: update uses table_id, delete does not. This avoids consuming an unset dataplane field and emitting invalid trace payloads.

 - About the LTTng issue:
 While working on this PR, I introduced a compilation issue related to the new tracepoints in zebra/interface.c during a rebase. Sorry about that — I was not familiar with LTTng tracing at the time. I am working on fixing this in the current series.

On the positive side, this led to adding CI coverage for --enable-lttng (thanks to @vjardin), which also uncovered an existing LTTng compilation issue already present on master. I fixed that master issue as well.
  
So I agree the diff is larger than the minimal vrf_id accessor change, but it is not arbitrary churn: it makes the delete/update requirements explicit and avoids relying on/tracing unset dataplane fields. The actual behavior change remains isolated in the second commit.

maxime-leroy commented:
@mjstapp  I added the extra {} blocks in interface_vrf_update() / interface_vrf_del()
in the first commit to preserve the original indentation and minimize the diff,
so it should be easier to review.

I’ve added a follow-up commit that removes these redundant blocks and runs
clang-format.

Hope this makes the review easier on your side.

maxime-leroy commented:
ci:rerun

maxime-leroy committed:
zebra: do not rely on dplane table_id for VRF delete

zebra_if_dplane_ifp_handling() was reading dplane_ctx_get_ifp_table_id()
for VRF events.

In the netlink dplane backend, ifp_table_id is only set via
netlink_vrf_change(), which is invoked from netlink_link_change() for
RTM_NEWLINK VRF events. It is not set for RTM_DELLINK. This causes
interface_vrf_change() to receive table_id as 0 on VRF deletes.

For delete operations, save the VRF pointer before if_delete_update()
and retrieve table_id from vrf->data.l.table_id. For add/update
operations, continue using the dplane-provided table_id.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>

maxime-leroy committed:
zebra: pass vrf pointer to interface_vrf_change for delete

In the delete path, we already hold a reference to the vrf structure
before calling if_delete_update(). Passing this pointer directly to
interface_vrf_change() avoids a redundant vrf_lookup_by_id() call and
allows accessing vrf->name, vrf->vrf_id, and vrf->data.l.table_id
directly.

For update operations, the vrf pointer is passed as NULL and the
function continues to use the vrf_id, name, and tableid parameters
from the dataplane context.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>

maxime-leroy commented:
ci:rerun

maxime-leroy commented:
@mjstapp any update? The patch is now the minimal version you asked  for (+30/-38) since January 27. 
                                                                                                                                                                                                                                  
FWIW, greptile (configured by @mwinter-osr on an unrelated FRR PR) also recommends splitting this kind of function:  https://github.com/opensourcerouting/frr/pull/234#discussion_r3079946094

mjstapp commented on zebra/zebra_trace.h:
```diff
@@ -91,9 +91,9 @@ TRACEPOINT_LOGLEVEL(frr_zebra, if_upd_ctx_dplane_result, TRACE_INFO)
 	frr_zebra,
 	if_vrf_change,
-	TP_ARGS(ifindex_t, ifindex, const char *, name, uint32_t, tableid, uint8_t, loc),
+	TP_ARGS(vrf_id_t, vrf_id, const char *, name, uint32_t, tableid, uint8_t, loc),
```
hmm - I don't think we can just change an existing trace output (without a deprecation cycle) - there may be existing users.
can the tracepoint be moved to a place where the arguments it uses are available?

maxime-leroy replied:
done

mjstapp commented on zebra/interface.c:
```diff
@@ -2013,16 +2000,19 @@ static void zebra_if_dplane_ifp_handling(struct zebra_dplane_ctx *ctx)
 		enum zebra_slave_iftype zif_slave_type;
 		uint8_t bypass;
 		uint64_t flags;
-		vrf_id_t vrf_id;
```
was there a reason to move this around - was its existing location a problem?

maxime-leroy replied:
vrf_id is now needed earlier in this block (passed into interface_vrf_change(op, vrf_id, ...)), so it's initialized at declaration instead of being assigned further down. 

Moving the line up to the second slot is purely cosmetic, to keep the reverse christmas tree ordering of the surrounding declarations.

maxime-leroy committed:
zebra: use dplane provided vrf_id instead of casting ifindex

interface_vrf_change() was implicitly assuming that the VRF netdevice
ifindex could be used as zebra's vrf_id.

This assumption holds true for the Linux kernel dataplane, where the VRF
ID is defined as the ifindex of the VRF interface, so this change does
not alter kernel behavior.

However, the dataplane API already exposes both concepts explicitly via
dplane_ctx_get_ifindex() and dplane_ctx_get_ifp_vrf_id(ctx). Using the
proper accessor avoids casting an ifindex to vrf_id_t and better respects
the dataplane abstraction.

On interface updates, the vrf_id provided by the dataplane is now used
directly. On interface deletion (DELLINK), where the dataplane context
may no longer carry vrf information, zebra relies on the existing ifp
state (ifp->vrf->vrf_id) before if_delete_update() is called. The table_id
is also retrieved from ifp->vrf->data.l.table_id since
dplane_ctx_get_ifp_table_id() is not set for RTM_DELLINK events.

The if_vrf_change LTTng tracepoint is moved out of interface_vrf_change()
and into the caller zebra_if_dplane_ifp_handling(), where the original
ifindex from the dplane context is still available. This preserves the
existing tracepoint ABI (field name "ifindex", type ifindex_t) so that
existing trace consumers are not affected.

This is required for non-kernel dataplanes such as Grout (DPDK), where
the VRF ID is not equal to the VRF interface ifindex. In that case, using
the correct vrf_id fixes VRF handling between zebra and the Grout
dataplane.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>

mjstapp reviewed (approved):
Thanks, looks good to me now

vjardin reviewed (approved):
Ok
